tsdb/engine/tsm1/engine.go

```diff
@@ -223,50 +228,92 @@ func (e *DevEngine) Begin(writable bool) (tsdb.Tx, error) {
 
 func (e *DevEngine) WriteTo(w io.Writer) (n int64, err error) { panic("not implemented") }
 
-func (e *DevEngine) compact() {
-	for {
+func (e *DevEngine) writeSnapshot() error {
+	// Lock and grab the cache snapshot along with all the closed WAL
+	// filenames associated with the snapshot
+	closedFiles, snapshot, compactor, err := func() ([]string, *Cache, *Compactor, error) {
+		e.mu.Lock()
+		defer e.mu.Unlock()
+
 		if err := e.WAL.CloseSegment(); err != nil {
-			e.logger.Printf("error rolling current WAL segment: %v", err)
-			time.Sleep(time.Second)
-			continue
+			return nil, nil, nil, err
 		}
 
-		tsmFiles, segments, err := e.CompactionPlan.Plan()
+		segments, err := e.WAL.ClosedSegments()
 		if err != nil {
-			e.logger.Printf("error calculating compaction plan: %v", err)
-			time.Sleep(time.Second)
-			continue
+			return nil, nil, nil, err
+		}
+
+		snapshot := e.Cache.Snapshot()
+
+		return segments, snapshot, e.Compactor.Clone(), nil
+	}()
+
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		// write the new snapshot files
+		newFiles, err := compactor.WriteSnapshot(snapshot)
+		if err != nil {
+			e.logger.Printf("error writing snapshot from compactor: %v", err)
+			return
+		}
+
+		e.mu.RLock()
```

Comment:
It was intentionally moved below the `WriteSnapshot` call so that concurrent writes could run.  As long as `e.logger` is never modified, this should be fine.